common/src/leap/soledad/common/couch.py

```diff
@@ -1474,14 +1506,32 @@ def ensure_database(self, dbname):
         databases. But, in case it ever does, by raising an exception here we
         have one more guarantee that no modified client will be able to
         enforce creation of a database when syncing.
+        This can only be done by using a shell script with proper validation
+        and permissions (create_cmd).
 
         :param dbname: The name of the database to ensure.
         :type dbname: str
 
         :raise Unauthorized: Always, because Soledad server is not allowed to
                              create databases.
+
+        :return: The CouchDatabase object and its replica_uid.
+        :rtype: (CouchDatabase, str)
         """
-        raise Unauthorized()
+        if not self.create_cmd:
+            raise Unauthorized()
+        else:
+            code, out = exec_validated_cmd(self.create_cmd, dbname,
+                                           validator=is_db_name_valid)
+            if code is not 0:
+                logger.error("""
+                    Error while creating database (%s) with (%s) command.
+                    Output: %s
+                    Exit code: %d
+                    """ % (dbname, self.create_cmd, out, code))
+                raise Unauthorized()
```

Comment:
During an outage or error, the client won't receive a replica_uid to save. That means it will trigger this code again and it is fine to run multiple times. This is why it is originally called "ensure", but the reason lies on a lot more code spread over client and server codebase.
This commit helps a bit to see this: https://github.com/pixelated/soledad/commit/3c725026c8f1c3b61b16e6c353dba2df078a76c0
It is based on original u1db behaviour. When it reaches a DatabaseDoesNotExist error, a callback to save the replica_uid of the recently create database will be passed to SyncTarget on client. That in turn will fire "ensure" as True on server side, making the server call this method (see sync.py). Then, the current version behaviour is just to throw it all up and present the error to user (nothing to do, since we got "unauthorized" back due to server lacking code to ensure a database). This PR adds this ability back. So, where you would be getting those errors, the server will now be attempting to create a new database for you.
About the Unauthorized raised, I decided to leave this as fallback because the logic there ensures that if something went wrong calling the command (misconfiguration, tampering or disabled feature) the previous behaviour is restored and Unauthorized is raised, leaving less clues for a possible attacker on what happened with the command.